rust/src/dataset.rs

```diff
@@ -471,6 +468,20 @@ impl Dataset {
         })
     }
 
+    /// Write to or Create a [Dataset] with a stream of [RecordBatch]s.
+    ///
+    /// Returns the newly created [`Dataset`]. Returns [Error] if the dataset already exists.
+    pub async fn write(
+        batches: impl RecordBatchReader + Send + 'static,
+        uri: &str,
+        params: Option<WriteParams>,
+    ) -> Result<Self> {
+        // Box it so we don't monomorphize for every one. We take the generic
+        // parameter for API ergonomics.
+        let batches = Box::new(batches);
+        Self::write_impl(batches, uri, params).await
```

Comment:
This trick is based on: 
https://nnethercote.github.io/perf-book/compile-times.html#llvm-ir